### ISSUE TITLE
Best approach for using transformed VEX code for analysis

### ISSUE THREAD

**Issue by mattrepl** [angr/angr#290]
I'd like to use some of angr's analyses on transformed VEX code. Some of the analyses I'm interested in are DFG and DDG generation which both require a CFG. It seems a reasonable approach to using transformed VEX code would be to first generate a CFG and then replace the node's IRSB instances with transformed IRSB instances. This modified CFG would then be used as input for analysis.

One snag, the current DFG implementation actually calls `angr.lifter.Lifter#lift` for each node rather than using the node's attached IRSB. That's an easy fix.

Any thoughts on how to best use angr analyses with transformed VEX code? Or comments about whether it's a terrible, out-of-scope idea? ;) 

Thanks!

**Comment by frapik99** [comment]
Hi @mattrepl ,
I'm sorry but I have just posted a similar just [https://github.com/angr/simuvex/issues/49](url). I didn't see your before...

My approch would be to create a new plugin once you have the path but I'm not sure of it

Thanks

**Comment by ltfish** [comment]
@mattrepl What DFG are you referring to?

**Comment by mattrepl** [comment]
@ltfish Line 56 in angr/analyses/dfg.py.

**Comment by ltfish** [comment]
WOW I'm surprised that the code got excavated... I think DDG is a better fit for your need. DFG is not maintained by any core developer in angr, and to be honest, I'm not sure that analysis is necessary to be kept in angr's analyses submodule.

**Comment by mattrepl** [comment]
Yeah, I agree that DDG would work. I was thinking of using DFG for now, since it's easier to reason about/modify and I believe DDG changes are happening internally?

Off-topic, but since we're talking about DDG and internal work being done, any idea when the new calling convention detection code will be made public? I'd like to do intra-function tracking of data, but that's difficult to do without knowing what's being passed in as arguments at call sites.

It may be the current version of DDG is usable for my situation, but I seem to recall it having problems last I looked at it a month ago. Will try again.

**Comment by mattrepl** [comment]
Just noticed DDG currently expects the project architecture to be x86 or amd64. Hopefully that's changing in the update.

Also, happy to help with this stuff. Just avoiding work on DDG until internal changes are released.

**Comment by zardus** [comment]
As a meta-comment: we're planning on moving all development to github (as opposed to developing on our internal gitlab and synchronizing things out periodically), so these sort of development bottlenecks will hopefully be more rare in the future.

In the meantime, we'll have to suffer a bit longer :-)

**Comment by mattrepl** [comment]
Can't wait! And thank you all for the work you put into the project and talking with the community. It's great having an open-source, shared core.

**Comment by ltfish** [comment]
> I was thinking of using DFG for now, since it's easier to reason about/modify and I believe DDG changes are happening internally?

Yes I'm refactoring DDG right now (to make it easier to reason about :-) ) It's been going forward slowly since I was seriously sick.

> any idea when the new calling convention detection code will be made public?

No accurate estimate, but I'll try to get it delivered during Christmas.

> Just noticed DDG currently expects the project architecture to be x86 or amd64.

I didn't even such a limit exists! IMO there is no road blocker that prevents DDG from working on other architectures, probably just some small implementation issues. I'll definitely check it out.

**Comment by frapik99** [comment]
@ltfish Thanks for that, I'm interested too into DDG. I would like to understand how to query it and extract the deps for the variables. Would it be possible?

Thanks

**Comment by axt** [comment]
@frapik99 
I've created some visualizations of the DDG, which maybe could help you to better understand the relations in the graphs: 
Check it here: 
  https://github.com/axt/angr-utils/tree/master/examples/conceptual#ddg-1
  https://github.com/axt/angr-utils/tree/master/examples/conceptual#ddg-2
  https://github.com/axt/angr-utils/tree/master/examples/plot_ddg

**Comment by frapik99** [comment]
@axt excellent example, i'll read your code but probably I'll wait for the next version of the DDG.

Thanks
